dev/core#4808 Phase out profile listings

mlutfy opened the pull request:
Overview
----------------------------------------

Begins the process of hiding some profile options that are rarely used and often used in risky ways:

- Visibility: public / admin
- Searchable
- ~~View Only (not really risky, but I don't know what it's for?)~~ Not removed based on feedback

It also does a few minor tweaks:

- Removes '?' in labels
- Shows the button at the top only (like in Manage Contribution pages, Events, etc)

I was very tempted to remove the "Preview" link for each individual field, but maybe I'll do that in AdminUI only. The only reason I didn't, is because it had a bigger visual impact than expected.

Issue: https://lab.civicrm.org/dev/core/-/issues/4808

Before
----------------------------------------

![image](https://github.com/user-attachments/assets/692dc7ce-70a3-4540-87c6-191e3b57c45f)

After
----------------------------------------

Gone:

![image](https://github.com/user-attachments/assets/cd2d67fe-b7f4-4140-a20f-340b0048d418)

and also:

![image](https://github.com/user-attachments/assets/d532c3d8-043b-4529-981b-3a2e2fddf12c)

Tweak to permissions:

![image](https://github.com/user-attachments/assets/3671ad2a-1b93-46a8-b8f7-c964ec8593d6)

Tweak to the Advanced Search:

![image](https://github.com/user-attachments/assets/0a453dbf-f719-4fe5-9c27-fb2892e1eb56)


Comments
--------------------------------------

I'm really not sure what the "profile listings and forms" permission is actually used for. It is checked only in 3 places, one of which seems somewhat relevant (but also combined to "profile create").

Todo
----------------------------------------

If the extension is disabled:

- [x] Upgrader: check the profile field settings to see if those features are used. If not, enable the extension.
- [x] Profile settings: "Used For", remove the "Directory" in "Standalone Form or Directory"?
- [x] Profile settings: remove: Advanced Settings > "Limit listings to a specific group?"
- [x] Profile settings: remove: Advanced Settings > "Include profile edit links in search results?" and "Include Standalone user account .." (only applies to listings)
- [x] Block listings in `CRM/Profile/Page/Listings.php` (that way, disabling the legacyprofiles extensions becomes an easy way of securing the site)
- [x] Profile listings permission: would make sense to move it to the extension, but doing so might cause Drupal to lose the permission (if the cache is flushed before the extension is enabled). So I am tempted to only adjust the permission label.

mlutfy committed:
dev/core#4808 Phase out public profile fields and profile listings

mlutfy commented:
A bit of cleanup in the Profile "Advanced Settings":

![image](https://github.com/user-attachments/assets/b45a222c-2515-4268-a14c-05ab5c5397e2)

kcristiano commented:
@mlutfy for the parameters that are removed how would an admin set these?

- Visibility: public / admin  is used to show fileds on a profile just to staff (admin) and not on the public facing profile.
- View Only - This is used on occasion to display data to the user/contact that they cannot change.  Would we need code ovrrides?

Profiles are not dead yet, we need them for contribution pages and event pages.  We also can only use profiles if we want contact image uploads as this is does not reliably work in Form Builder.

I wonder if this is just a little early to remove functionality we don't have a replacement for yet, or maybe I am missing what problem this solves.

mlutfy commented:
More tweaks, when viewing the Profile fields:

![image](https://github.com/user-attachments/assets/bbfa1a6c-3472-481f-8f6e-f625285578e2)

kcristiano commented:
> My focus in on removing Profile Listings, and a bit on removing "Profile View", which I think are truly evil, and I'm trying to do a few improvements along the way.

While I agree we still have some of these in use.   If I understand correctly, your intent is to make it so new listings could not be created.

> Visibility: my understanding is that, unlike PriceSets, this is only for Profile Listings / View?

It does work and somewhat useful in edit mode.   It is less useful than Price sets for certain.

mlutfy commented:
> If I understand correctly, your intent is to make it so new listings could not be created.

Yep - hidden for new installs. Or from installs that never used them. And it can still be enabled.

SearchKit is a good replacement for this.

> It does work and somewhat useful in edit mode.

How does it work? It lets non-admins edit any contact?

Here is something I often see on sites:

![image](https://github.com/user-attachments/assets/14e722f6-06ad-4da2-8260-4c2f13400196)

If the profile listing permission is enabled, and they enable the "public" visibility setting, then they have a security problem.

kcristiano commented:
To answer on View only -

It makes a field on a profile View only - say the org has a custom contact field that the set and do not want changed.  They can expose it on the profile edit screen, the contact can see and validate but not change the information.   So Very limited use case.   

I  agree about the profile listings capability and prefer Form Builder going forward.  

I'll give this a test as soon as I can, but I think this is a good improvement.  Thank you for clarifying for me.

shaneonabike commented:
This is great! Personally, I would either make the Dedupe required OR you don't select anything. I have several big clients presently who are doing a massive cleanup after multiple contacts have been created, and they are having an issue keeping up with that. Setting the Dedupe rule to what you set already could potentially cause someone to setup a profile and be unaware of the dedupe (unfortunately even though it's right there loads of people miss it)

mlutfy commented:
@shaneonabike Thanks for the feedback. You're right, this is a really important option, but it only applies to Profiles set in "Standalone Form" mode. I changed the default from "do nothing" to "create duplicates", because I've seen that on client sites too, and it's really bad. For example, people might fail to signup to a newsletter, staff often don't receive the complaints from confused visitors.

One of the follow-up changes I would like to do, is to convert the "Used For" checkboxes to a single Select (if possible), and then I think the Settings all only apply to either Forms/Listings, so they should only be displayed only in those circumstances.

In an ideal  world, the Profiles should require a dedupe rule, instead of asking how we want to dedupe (and the 3 options are all rather bad). That's a rather  big change though, for something considered legacy, so I will probably not go that far. Hopefully one day FormBuilder will make it possible to signup to a newsletter, or at least, better to invest energy there :grimacing:

mlutfy commented:
I updated the labels for the Advanced Search (Display Columns) in a previous comment:

https://github.com/civicrm/civicrm-core/pull/32648#issuecomment-2831113684

based on [feedback](https://chat.civicrm.org/civicrm/pl/3bpnoey1p3gj7gh1x3u66xf4eo) by @petednz 

I also did a minor update to a profile listing permission description (screenshot in the PR summary at the top).

and moved `CRM/Profile/Page/{Listings,View}.php` to the `ext/legacyprofiles` directory.

I think that the last TODO is to figure out the conditions in the Upgrader for enabling the extension.

mlutfy commented:
All done - ready for review :tada:

mlutfy commented:
rebased

mlutfy commented:
@colemanw Fixed - and tests are passing.

mlutfy commented on CRM/Upgrade/Incremental/php/SixFive.php:
```diff
@@ -29,6 +29,38 @@ class CRM_Upgrade_Incremental_php_SixFive extends CRM_Upgrade_Incremental_Base {
+  public static function installLegacyProfiles(): bool {
+    // Based on the instructions for the FiveThirty financialacls upgrade step
+    // Install via direct SQL manipulation. Note that:
+    // (1) This extension has no activation logic as of 5.75 (the DB tables are still in core)
```
Fixed (I almost feel bad for the test servers) :)

mlutfy commented:
Thanks! I'll be deploying this patch to clients, so hopefully we'll catch any glitches.
